File: internal/cli/snapshot.go
package cli

import (
	"fmt"
	"strings"

	"github.com/qyinm/gandalf/internal/gandalfcore/snapshot"
	"github.com/qyinm/gandalf/internal/gandalfcore/store"
	"github.com/qyinm/gandalf/internal/gandalfcore/types"
	"github.com/spf13/cobra"
)

func newSnapshotCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "snapshot",
		Short: "Create, list, and show snapshots",
	}
	cmd.AddCommand(newSnapshotCreateCmd())
	cmd.AddCommand(newSnapshotListCmd())
	cmd.AddCommand(newSnapshotShowCmd())
	return cmd
}

func newSnapshotCreateCmd() *cobra.Command {
	var common CommonFlags
	var name string
	var metadataOnly bool

	cmd := &cobra.Command{
		Use:   "create",
		Short: "Capture and persist a snapshot",
		RunE: func(cmd *cobra.Command, _ []string) error {
			exitCode := runSnapshotCreate(cmd, &common, name, metadataOnly)
			if exitCode != 0 {
				return errExit(exitCode)
			}
			return nil
		},
	}

	common.bindFlags(cmd.Flags())
	cmd.Flags().StringVar(&name, "name", "", "Snapshot name")
	_ = cmd.MarkFlagRequired("name")
	cmd.Flags().BoolVar(&metadataOnly, "metadata-only", false, "Store metadata only (no content capture)")
	return cmd
}

func newSnapshotListCmd() *cobra.Command {
	var common CommonFlags

	cmd := &cobra.Command{
		Use:   "list",
		Short: "List snapshots in the store",
		RunE: func(cmd *cobra.Command, _ []string) error {
			exitCode := runSnapshotList(cmd, &common)
			if exitCode != 0 {
				return errExit(exitCode)
			}
			return nil
		},
	}

	common.bindFlags(cmd.Flags())
	return cmd
}

func newSnapshotShowCmd() *cobra.Command {
	var common CommonFlags

	cmd := &cobra.Command{
		Use:   "show [name]",
		Short: "Show snapshot metadata (or full JSON with --json)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			exitCode := runSnapshotShow(cmd, &common, args[0])
			if exitCode != 0 {
				return errExit(exitCode)
			}
			return nil
		},
	}

	common.bindFlags(cmd.Flags())
	return cmd
}

func runSnapshotCreate(cmd *cobra.Command, common *CommonFlags, name string, metadataOnly bool) int {
	runtime, snapErr := resolveRuntime(common)
	if snapErr != nil {
		return writeError(cmd.ErrOrStderr(), snapErr)
	}

	contentBackedCodexUser := runtime.Agent != nil &&
		*runtime.Agent == types.AgentCodex &&
		runtime.Scope != nil &&
		*runtime.Scope == types.ScopeUser

	if !metadataOnly && !contentBackedCodexUser {
		return writeError(cmd.ErrOrStderr(), &types.SnapError{
			Code:    "GANDALF_METADATA_ONLY_REQUIRED",
			Problem: "Snapshots are metadata-only.",
			Cause:   "`snapshot create` was called without `--metadata-only`.",
			Fix:     "Add `--metadata-only`, or use `--agent codex --scope user` for the Codex rollback safety-net path.",
		})
	}

	captureRuntime := runtime
	captureRuntime.CaptureContent = !metadataOnly && contentBackedCodexUser

	state, err := snapshot.CaptureCurrentState(&captureRuntime, name)
	if err != nil {
		return writeError(cmd.ErrOrStderr(), &types.SnapError{
			Code:    "GANDALF_SNAPSHOT_CAPTURE_FAILED",
			Problem: "Failed to capture current state.",
			Cause:   err.Error(),
			Fix:     "Verify project and store paths are accessible.",
		})
	}

	if err := store.WriteSnapshot(runtime.StoreDir, store.StoreSnapshotFrom(state.Snapshot), runtime.Agent); err != nil {
		return writeError(cmd.ErrOrStderr(), &types.SnapError{
			Code:    "GANDALF_SNAPSHOT_WRITE_FAILED",
			Problem: "Failed to write snapshot.",
			Cause:   err.Error(),
			Fix:     "Verify the store directory is writable.",
		})
	}

	kind := "content-backed"
	if metadataOnly {
		kind = "metadata-only"
	}
	line := fmt.Sprintf("Created %s snapshot: %s", kind, name)
	if runtime.Agent != nil {
		line += fmt.Sprintf(" (agent: %s)", runtime.Agent.String())
	}
	if runtime.Scope != nil {
		line += fmt.Sprintf(" (scope: %s)", runtime.Scope.String())
	}
	line += "\n"
	return writeStdout(cmd.OutOrStdout(), line)
}

func runSnapshotList(cmd *cobra.Command, common *CommonFlags) int {
	runtime, snapErr := resolveRuntime(common)
	if snapErr != nil {
		return writeError(cmd.ErrOrStderr(), snapErr)
	}

	names, err := store.ListSnapshots(runtime.StoreDir, runtime.Agent)
	if err != nil {
		return writeError(cmd.ErrOrStderr(), &types.SnapError{
			Code:    "GANDALF_SNAPSHOT_LIST_FAILED",
			Problem: "Failed to list snapshots.",
			Cause:   err.Error(),
			Fix:     "Verify the store directory exists and is readable.",
		})
	}

	if common.JSON {
		return writeJSON(cmd.OutOrStdout(), names)
	}

	if len(names) == 0 {
		return writeStdout(cmd.OutOrStdout(), "No snapshots.\n")
	}
	return writeStdout(cmd.OutOrStdout(), strings.Join(names, "\n")+"\n")
}

func runSnapshotShow(cmd *cobra.Command, common *CommonFlags, name string) int {
	runtime, snapErr := resolveRuntime(common)
	if snapErr != nil {
		return writeError(cmd.ErrOrStderr(), snapErr)
	}

	snap, err := store.ReadSnapshot(runtime.StoreDir, name, runtime.Agent)
	if err != nil {
		return writeError(cmd.ErrOrStderr(), &types.SnapError{
			Code:    "GANDALF_SNAPSHOT_NOT_FOUND",
			Problem: fmt.Sprintf("Snapshot %q not found.", name),
			Cause:   err.Error(),
			Fix:     "Run `gandalf snapshot list` to see available snapshots.",
		})
	}

	if common.JSON {
		return writeJSON(cmd.OutOrStdout(), snap)
	}
	return writeStdout(cmd.OutOrStdout(), snap.Manifest.Name+"\n")
}

File: apps/cli/src/commands/snapshot.ts
/**
 * snapshot command: create, list, show snapshots.
 *
 * Subcommands:
 *   snapshot create --name <name> --metadata-only --project .
 *   snapshot create --name <name> --agent codex --scope user --project .
 *   snapshot list
 *   snapshot show <name>  show <name> [--json]
 */

import { formatSnapError } from "@qxinm/gandalf-core/errors.js";
import { captureTimelineSnapshot } from "@qxinm/gandalf-core/timeline.js";
import {
  listSnapshots,
  readSnapshot,
} from "@qxinm/gandalf-core/store.js";
import React from "react";
import { hasFlag, json, valueAfter } from "../cli-shared.js";
import { detectTuiMode, isInkMode, renderComponent } from "@qxinm/gandalf-tui";
import type { Command, CommandContext } from "./index.js";

/* ------------------------------------------------------------------ */
/*  Command definition                                                 */
/* ------------------------------------------------------------------ */

export const snapshotCommand: Command = {
  name: "snapshot",
  description: "Create, list, and show snapshots.",

  async execute(ctx: CommandContext): Promise<number> {
    const { args, options } = ctx;
    const sub = args[1];

    /* ---------- snapshot create ---------- */
    if (sub === "create") {
      // --tui: interactive wizard
      const tuiOpts = detectTuiMode(args);
      if (tuiOpts.mode !== "none") {
        const { snapshotCreateWizard } = await import("@qxinm/gandalf-tui/wizards/snapshot-create.js");
        return snapshotCreateWizard(options);
      }
      const name = valueAfter(args, "--name");
      if (!name) {
        process.stderr.write(
          formatSnapError({
            code: "GANDALF_MISSING_NAME",
            problem: "Snapshot name is required.",
            cause: "`snapshot create` was called without `--name`.",
            fix: "Run `gandalf snapshot create --name baseline --metadata-only --project .`."
          })
        );
        return 1;
      }
      const metadataOnly = hasFlag(args, "--metadata-only");
      const contentBackedCodexUser = options.agent === "codex" && options.scope === "user";
      if (!metadataOnly && !contentBackedCodexUser) {
        process.stderr.write(
          formatSnapError({
            code: "GANDALF_METADATA_ONLY_REQUIRED",
            problem: "Snapshots are metadata-only.",
            cause: "`snapshot create` was called without `--metadata-only`.",
            fix: "Add `--metadata-only`, or use `--agent codex --scope user` for the Codex rollback safety-net path."
          })
        );
        return 1;
      }

      await captureTimelineSnapshot({
        ...options,
        captureContent: !metadataOnly && contentBackedCodexUser
      }, {
        snapshotName: name,
        title: name
      });
      process.stdout.write(`Created ${metadataOnly ? "metadata-only" : "content-backed"} snapshot: ${name}`);
      if (options.agent) process.stdout.write(` (agent: ${options.agent})`);
      if (options.scope) process.stdout.write(` (scope: ${options.scope})`);
      process.stdout.write("\n");
      return 0;
    }

    /* ---------- snapshot list ---------- */
    if (sub === "list") {
      const names = await listSnapshots(options.storeDir, options.agent);
      if (isInkMode(args)) {
        const { default: SnapshotList } = await import("@qxinm/gandalf-tui/components/SnapshotList.js");
        return renderComponent(
          () => React.createElement(SnapshotList, { names })
        );
      }
      process.stdout.write(names.length === 0
        ? "No snapshots.\n"
        : `${names.join("\n")}\n`);
      return 0;
    }

    /* ---------- snapshot show ---------- */
    if (sub === "show") {
      const name = args[2];
      if (!name) {
        process.stderr.write(
          formatSnapError({
            code: "GANDALF_MISSING_NAME",
            problem: "Snapshot name is required.",
            cause: "`snapshot show` was called without a name.",
            fix: "Run `gandalf snapshot list` and pass one of the listed names."
          })
        );
        return 1;
      }
      const snapshot = await readSnapshot(options.storeDir, name, options.agent);
      process.stdout.write(hasFlag(args, "--json") ? json(snapshot) : `${snapshot.manifest.name}\n`);
      return 0;
    }

    /* ---------- unknown subcommand ---------- */
    process.stderr.write(
      formatSnapError({
        code: "GANDALF_UNKNOWN_SUBCOMMAND",
        problem: `Unknown snapshot subcommand: "${sub ?? ""}".`,
        cause: "`snapshot` was called with an unrecognized subcommand.",
        fix: "Use `create`, `list`, or `show`. Run `gandalf --help` for details."
      })
    );
    return 1;
  }
};

export default snapshotCommand;
